### ISSUE TITLE
 GOPATH set to GOROOT has no effect 

### ISSUE THREAD

**Issue by 3052** [golang/go#70093]
### Go version

go version go1.23.2 windows/amd64

### Output of `go env` in your module/workspace:

```shell
set GO111MODULE=
set GOARCH=amd64
set GOBIN=
set GOCACHE=C:\Users\spenny\AppData\Local\go-build
set GOENV=C:\Users\spenny\AppData\Roaming\go\env
set GOEXE=.exe
set GOEXPERIMENT=
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=C:\Users\spenny\go\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=C:\Users\spenny\go
set GOPRIVATE=
set GOPROXY=https://proxy.golang.org,direct
set GOROOT=C:\Users\spenny\go
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLCHAIN=auto
set GOTOOLDIR=C:\Users\spenny\go\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.23.2
set GODEBUG=
set GOTELEMETRY=local
set GOTELEMETRYDIR=C:\Users\spenny\AppData\Roaming\go\telemetry
set GCCGO=gccgo
set GOAMD64=v1
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=0
set GOMOD=NUL
set GOWORK=
set CGO_CFLAGS=-O2 -g
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-O2 -g
set CGO_FFLAGS=-O2 -g
set CGO_LDFLAGS=-O2 -g
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -fno-caret-diagnostics -Qunused-arguments -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=C:\Users\spenny\AppData\Local\Temp\go-build4205015363=/tmp/go-build -gno-record-gcc-switches
```

### What did you do?

repost because previous issue was closed in error

go version

### What did you see happen?

warning: GOPATH set to GOROOT (C:\Users\spenny\go) has no effect
go version go1.23.2 windows/amd64

### What did you expect to see?

go version go1.23.2 windows/amd64

note I also tried to unset both values:

PS C:\Users\spenny\Desktop> go env -u GOROOT
warning: GOPATH set to GOROOT (C:\Users\spenny\go) has no effect
PS C:\Users\spenny\Desktop> go env -u GOPATH
warning: GOPATH set to GOROOT (C:\Users\spenny\go) has no effect
PS C:\Users\spenny\Desktop> go version
warning: GOPATH set to GOROOT (C:\Users\spenny\go) has no effect
go version go1.23.2 windows/amd64

also:

PS C:\Users\spenny\Desktop> go env -changed
warning: GOPATH set to GOROOT (C:\Users\spenny\go) has no effect

**Comment by ianlancetaylor** [comment]
The default for `GOPATH` is the directory `go` in your home directory.  You have installed Go such that that is also your `GOROOT`.  If you want to use that value for `GOROOT`, you must explicitly set `GOPATH` to some other directory.

I'm going to close this because there is nothing to change in Go.  If you want to discuss this, please use a forum; see https://go.dev/wiki/Questions.  Thanks.

**Comment by 3052** [comment]
@ianlancetaylor  this comment is extremely salient:

https://github.com/golang/go/issues/65656#issuecomment-2440134469

are you really not willing to admit that the current state is confusing?

**Comment by ianlancetaylor** [comment]
I certainly agree that the current state is confusing.

What should we do about it?

If your goal with this issue is to get a better error message, I would encourage you to state that explicitly.  This bug report doesn't say what the problem is and it doesn't make any suggestion as to what would be better.  It is quite possible that these are very obvious to you.  That does not mean that they are obvious to us.  Thanks.

**Comment by 3052** [comment]
Why would I do that? You've made it clear you're more interested in shutting down discussion instead of fostering it. If you mean what you say, reopen the issue and I will help however I can

**Comment by ianlancetaylor** [comment]
You are absolutely correct about discussion: we've found that discussion on the issue tracker does not work.  That's why we refer people to https://go.dev/wiki/Questions.  The issue tracker is for bug reports.  A bug report is a two-way street: we have to be able to understand what the problem is, and you have to take the time to explain the problem.  It would be nice if we could just figure it out, but we can't.  We are tiny team supporting millions of users.  Help us out.

**Comment by 3052** [comment]
the current summary by @tonymet does a great job in summarizing the current problem:

https://github.com/golang/go/issues/65656#issuecomment-2440134469

in my opinion, nothing needs to be added to that. if someone need further detail I am happy to provide it. in short:

some users are gonna get this warning and have no idea how to fix it, and the Go tool not only takes not action to help the problem, it doesn't even advise on an action. a simple "move your Go installation to another folder" or similar would go a long way

**Comment by xieyuschen** [comment]
[Go Wiki: Setting GOPATH](https://go.dev/wiki/SettingGOPATH) has a note:

> Note that GOPATH must not be the same path as your Go installation.

However, it doesn't say what will happen when they're the same so it's hard for users to refer this note because there is no link between warning _warning: GOPATH set to GOROOT (%s) has no effect_ and this go wiki part.

We can consider to change 2 things:
### compiler message:
The comment in L151 probably is better because user could easily know the behavior of _GOPATH is not set_ by searching the GOPATH doc. 
https://github.com/golang/go/blob/67f131485541f362c8e932cd254982a8ad2cfc09/src/cmd/go/main.go#L150-L153
User could reach the go wiki about GOPATH and find the default behavior of GOPATH, and they can easily come up with a solution by specifying GOPATH to another place. 
>  If no GOPATH is set, it is assumed to be $HOME/go on Unix systems and %USERPROFILE%\go on Windows.

```diff
// main.go:
-		fmt.Fprintf(os.Stderr, "warning: GOPATH set to GOROOT (%s) has no effect\n", gopath)
+		fmt.Fprintf(os.Stderr, "warning: GOPATH set to GOROOT (%s) has no effect and it is equivalent to not setting GOPATH at all \n", gopath)

```

### go wiki:  

After changing the go error message, we can consider to add more details for _Note that GOPATH must not be the same path as your Go installation._

I prefer to add 2 parts:

1. Put the compiler warning message so it ensures the user could search it from search engine quickly.
2. Add a sentence to guide users how to solve it if "the GOROOT and GOPATH are the same for some reasons".  

```diff
// go wiki
Note that GOPATH must not be the same path as your Go installation.
+Consider to set GOPATH to another place which differs from GOROOT
+if you encountered a warning "GOPATH set to GOROOT...".
```
In this user case, users GOROOT and GOPATH are the same, so go compiler will try to use `%USERPROFILE%\go` to fallback. However, after fallback the GOPATH still equals to the GOROOT because the GOROOT is in the fallback place.
Instead of ending up with _must not be the same_ in go doc, we can add some sentences so users could understand how to solve it.

HDYT? @ianlancetaylor @3052 If you think that's reasonable, I will send a CL in go. But need @ianlancetaylor help to amend go wiki because it requires the member role in golang organization and I'm not a member:(

**Comment by 3052** [comment]
+		fmt.Fprintf(os.Stderr, "warning: GOPATH set to GOROOT (%s) has no effect and it is equivalent to not setting GOPATH at all \n", gopath)

I dont think this goes far enough. the root cause here is the user has put their Go installation in the wrong place, so the message should reflect that. if user is determined then the fix could be messing with `go env`, but I think in most cases the correct fix is just moving the `go` folder somewhere else.

**Comment by ianlancetaylor** [comment]
I improved https://go.dev/wiki/InstallTroubleshooting, and sent https://go.dev/cl/623815 to change the error message to point to that wiki page.
